huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_black_white_list.go

```diff
@@ -0,0 +1,237 @@
+			},
+			"white_ip_list": {
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Optional:    true,
+				Computed:    true,
+				Description: `Specifies the white ip list. The minimum length is 7. The maximum length is 128.`,
+			},
+		},
+	}
+}
+
+func resourceBlackWhiteListCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.NewServiceClient("aad", "")
+	if err != nil {
+		return diag.Errorf("error creating CNAD Client: %s", err)
+	}
+
+	policyID := d.Get("policy_id").(string)
+	blackIpList := utils.ExpandToStringList(d.Get("black_ip_list").([]interface{}))
+	whiteIpList := utils.ExpandToStringList(d.Get("white_ip_list").([]interface{}))
+
+	if err := addPolicyIpList(client, blackIpList, "black", policyID); err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := addPolicyIpList(client, whiteIpList, "white", policyID); err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(policyID)
+	return resourceBlackWhiteListRead(ctx, d, meta)
+}
+
+func resourceBlackWhiteListRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	var mErr *multierror.Error
+
+	var (
+		getBlackWhiteListHttpUrl = "v1/cnad/policies/{policy_id}"
+		getBlackWhiteListProduct = "aad"
+	)
+	getBlackWhiteListClient, err := cfg.NewServiceClient(getBlackWhiteListProduct, "")
+	if err != nil {
+		return diag.Errorf("error creating CNAD Client: %s", err)
+	}
+
+	getBlackWhiteListPath := getBlackWhiteListClient.Endpoint + getBlackWhiteListHttpUrl
+	getBlackWhiteListPath = strings.ReplaceAll(getBlackWhiteListPath, "{policy_id}", d.Id())
+	getBlackWhiteListOpt := golangsdk.RequestOpts{
+		MoreHeaders: map[string]string{
+			"Content-Type": "application/json;charset=utf8",
+		},
+		KeepResponseBody: true,
+	}
+
+	getBlackWhiteListResp, err := getBlackWhiteListClient.Request("GET", getBlackWhiteListPath,
+		&getBlackWhiteListOpt)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving CNADAdvancedBlackWhiteList")
+	}
+
+	respBody, err := utils.FlattenResponse(getBlackWhiteListResp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	mErr = multierror.Append(
+		mErr,
+		d.Set("black_ip_list", utils.PathSearch("pop_policy.bw_list.black_ip_list",
+			respBody, nil)),
+		d.Set("white_ip_list", utils.PathSearch("pop_policy.bw_list.white_ip_list",
+			respBody, nil)),
+		d.Set("policy_id", d.Id()),
+	)
+
+	return diag.FromErr(mErr.ErrorOrNil())
+}
+
+func resourceBlackWhiteListUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	client, err := cfg.NewServiceClient("aad", "")
+	if err != nil {
+		return diag.Errorf("error creating CNAD Client: %s", err)
+	}
+
+	if d.HasChanges("black_ip_list") {
+		oldRaws, newRaws := d.GetChange("black_ip_list")
```

Comment:
how about change the type to schema.Set, the addIpList and deleteIpList can be get by the function of Difference and the unchanged ips will not be delete and add again

Reply:
OK, I have change it to `TypeSet`